server/server.go

```diff
@@ -0,0 +1,130 @@
+	p, err := strconv.ParseUint(req.GetPermission(), 10, 32)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Unable to parse permissions: %s", req.GetPermission()))
+
+	}
+
+	params := &config.MountParams{
+		Attributes:  req.GetAttributes(),
+		KubeSecrets: req.GetSecrets(),
+		TargetPath:  req.GetTargetPath(),
+		Permissions: os.FileMode(p),
+	}
+
+	cfg, err := config.Parse(params)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	smOpts := []option.ClientOption{option.WithUserAgent(s.UA)}
+
+	// Build the workload identity auth token if possible (fallback to node ID)
+	token, err := auth.Token(ctx, cfg, s.Kubeconfig)
+	if err != nil {
+		log.Printf("unable to use workload identity: %v", err)
+	} else {
+		smOpts = append(smOpts, option.WithTokenSource(oauth2.StaticTokenSource(token)))
+	}
+
+	// Build the secret manager client
+	client, err := secretmanager.NewClient(ctx, smOpts...)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to create secretmanager client: %v", err))
+	}
+
+	// Fetch the secrets from the secretmanager API and write them to the
+	// filesystem based on the SecretProviderClass configuration.
+	ovs, err := handleMountEvent(ctx, client, cfg)
+
+	return &v1alpha1.MountResponse{
+		ObjectVersion: ovs,
+	}, err
+}
+
+// Version implements provider csi-provider method
+func (s *Server) Version(ctx context.Context, req *v1alpha1.VersionRequest) (*v1alpha1.VersionResponse, error) {
+	return &v1alpha1.VersionResponse{
+		Version:        "v1alpha1",
+		RuntimeName:    "secrets-store-csi-driver-provider-gcp",
+		RuntimeVersion: s.RuntimeVersion,
+	}, nil
+}
+
+// handleMountEvent fetches the secrets from the secretmanager API and
+// writes them to the filesystem based on the SecretProviderClass configuration.
+func handleMountEvent(ctx context.Context, client *secretmanager.Client, cfg *config.MountConfig) ([]*v1alpha1.ObjectVersion, error) {
+	ovs := []*v1alpha1.ObjectVersion{}
+	for _, secret := range cfg.Secrets {
+		req := &secretmanagerpb.AccessSecretVersionRequest{
+			Name: secret.ResourceName,
+		}
+
+		result, err := client.AccessSecretVersion(ctx, req)
+		if err != nil {
+			// TODO: determine error codes, should we propagate error code space
+			// from the secret call to this response?
+			log.Printf("failed to access secret version (%s): %s", secret.ResourceName, err)
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		if err := ioutil.WriteFile(filepath.Join(cfg.TargetPath, secret.FileName), result.Payload.Data, cfg.Permissions); err != nil {
+			return nil, status.Error(codes.Internal, fmt.Sprintf("failed to write %s at %s: %s", secret.ResourceName, cfg.TargetPath, err))
+		}
+
+		log.Printf("secrets-store csi driver wrote %s at %s", secret.ResourceName, cfg.TargetPath)
+
+		ovs = append(ovs, &v1alpha1.ObjectVersion{
+			Id:      result.GetName(),
+			Version: result.GetName(),
```

Comment:
Should one of these be the Secret name and the other the version name?

Reply:
Verified that I believe that `Id` is best to be what the user requested in the `SecretProviderClass` and that `Version` should be the resolved version.